{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def dna_2_prot(f1, f2=\"translated_fasta.txt\"):\n",
    "    standard_code = {\n",
    "     \"UUU\": \"F\", \"UUC\": \"F\", \"UUA\": \"L\", \"UUG\": \"L\", \"UCU\": \"S\",\n",
    "     \"UCC\": \"S\", \"UCA\": \"S\", \"UCG\": \"S\", \"UAU\": \"Y\", \"UAC\": \"Y\",\n",
    "     \"UAA\": \"*\", \"UAG\": \"*\", \"UGA\": \"*\", \"UGU\": \"C\", \"UGC\": \"C\",\n",
    "     \"UGG\": \"W\", \"CUU\": \"L\", \"CUC\": \"L\", \"CUA\": \"L\", \"CUG\": \"L\",\n",
    "     \"CCU\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\", \"CAU\": \"H\",\n",
    "     \"CAC\": \"H\", \"CAA\": \"Q\", \"CAG\": \"Q\", \"CGU\": \"R\", \"CGC\": \"R\",\n",
    "     \"CGA\": \"R\", \"CGG\": \"R\", \"AUU\": \"I\", \"AUC\": \"I\", \"AUA\": \"I\",\n",
    "     \"AUG\": \"M\", \"ACU\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "     \"AAU\": \"N\", \"AAC\": \"N\", \"AAA\": \"K\", \"AAG\": \"K\", \"AGU\": \"S\",\n",
    "     \"AGC\": \"S\", \"AGA\": \"R\", \"AGG\": \"R\", \"GUU\": \"V\", \"GUC\": \"V\",\n",
    "     \"GUA\": \"V\", \"GUG\": \"V\", \"GCU\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\",\n",
    "     \"GCG\": \"A\", \"GAU\": \"D\", \"GAC\": \"D\", \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "     \"GGU\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\"}\n",
    "   \n",
    "    #import regular expression operations\n",
    "    import re\n",
    "    with open(f1) as file:\n",
    "        seq = [i.strip() for i in file]\n",
    "        #seperate and clean file name and sequence\n",
    "        while '' in file:\n",
    "            seq.remove('')\n",
    "    seqList = [i for i in seq]\n",
    "    #slice list to create new lists with just seq names and seq\n",
    "    namesOnly = seqList[0::2]\n",
    "    seqOnly = seqList[1::2]\n",
    "    seqDict = dict(zip(namesOnly, seqOnly))\n",
    "    #define table header and table column headings\n",
    "    with open(f2, 'w') as output:\n",
    "        rnaDict={key: value.upper().strip().replace(\"T\", \"U\") for key, value in seqDict.items()}\n",
    "        stopCodons=[\"UGA\", \"UAA\", \"UAG\"]\n",
    "        #Define empty string\n",
    "        prot=\"\"\n",
    "        #Start position is 0, end position is sequence length, increment is every three nucleotides\n",
    "        x=3\n",
    "        q=0\n",
    "        codonList = []\n",
    "        rnaItem = list(rnaDict.values())\n",
    "        for item in rnaItem:\n",
    "            for m in range(0,len(item),x):\n",
    "                codon=item[m:m+x]\n",
    "                if len(codon) == x:\n",
    "                    codonList.append(codon)   \n",
    "            #Define triplet as sequestered item in codonList\n",
    "        x=1\n",
    "        y=0\n",
    "        for triplet in codonList:\n",
    "            #Define string \"protein\" as the sequential \"\"+=\"\" value pairs of dictionary key imputs, triplets   \n",
    "            prot+=standard_code[str(triplet)]\n",
    "            x=x+1\n",
    "        amount = int(len(prot)/len(rnaItem))\n",
    "        split = amount\n",
    "        protList = [prot[i:i + split] for i in range(0, len(prot), split)]\n",
    "        protDict = dict(zip(namesOnly, protList))  \n",
    "        for k, v in protDict.items():\n",
    "            output.write(str(k) + \"\\n\" + str(v) + \"\\n\")\n",
    "        output.close()\n",
    "    return\n",
    "dna_2_prot(\"testDNA.txt\")\n",
    "\n",
    "seq1=\"AGCT\"\n",
    "seq2=\"AATTT\"\n",
    "def hot_code_DNA(seq):\n",
    "    standard_code = {\n",
    "     \"UUU\": \"F\", \"UUC\": \"F\", \"UUA\": \"L\", \"UUG\": \"L\", \"UCU\": \"S\",\n",
    "     \"UCC\": \"S\", \"UCA\": \"S\", \"UCG\": \"S\", \"UAU\": \"Y\", \"UAC\": \"Y\",\n",
    "     \"UAA\": \"*\", \"UAG\": \"*\", \"UGA\": \"*\", \"UGU\": \"C\", \"UGC\": \"C\",\n",
    "     \"UGG\": \"W\", \"CUU\": \"L\", \"CUC\": \"L\", \"CUA\": \"L\", \"CUG\": \"L\",\n",
    "     \"CCU\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\", \"CAU\": \"H\",\n",
    "     \"CAC\": \"H\", \"CAA\": \"Q\", \"CAG\": \"Q\", \"CGU\": \"R\", \"CGC\": \"R\",\n",
    "     \"CGA\": \"R\", \"CGG\": \"R\", \"AUU\": \"I\", \"AUC\": \"I\", \"AUA\": \"I\",\n",
    "     \"AUG\": \"M\", \"ACU\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "     \"AAU\": \"N\", \"AAC\": \"N\", \"AAA\": \"K\", \"AAG\": \"K\", \"AGU\": \"S\",\n",
    "     \"AGC\": \"S\", \"AGA\": \"R\", \"AGG\": \"R\", \"GUU\": \"V\", \"GUC\": \"V\",\n",
    "     \"GUA\": \"V\", \"GUG\": \"V\", \"GCU\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\",\n",
    "     \"GCG\": \"A\", \"GAU\": \"D\", \"GAC\": \"D\", \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "     \"GGU\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\"}\n",
    "    \n",
    "    output=[]\n",
    "    for i in seq:\n",
    "        if i == \"A\":\n",
    "            output.append([1,0,0,0])\n",
    "        if i == \"G\":\n",
    "            output.append([0,1,0,0])\n",
    "        if i == \"C\":\n",
    "            output.append([0,0,1,0])\n",
    "        if i == \"T\":\n",
    "            output.append([0,0,0,1])\n",
    "    print(output)\n",
    "    return\n",
    "\n",
    "def motif_finder(f1, motif, f2 = 'motif.txt'):\n",
    "    #set motif counter to 0\n",
    "    motifHits=0\n",
    "    #import regular expression operations\n",
    "    import re\n",
    "    with open(f1) as file:\n",
    "        seq = [i.strip() for i in file]\n",
    "        #seperate and clean file name and sequence\n",
    "        while '' in file:\n",
    "            seq.remove('')\n",
    "    seqList = [i.strip(\">\") for i in seq]\n",
    "    #slice list to create new lists with just seq names and seq\n",
    "    namesOnly = seqList[0::2]\n",
    "    seqOnly = seqList[1::2]\n",
    "    #define table header and table column headings\n",
    "    col1=\"SeqName\"\n",
    "    col2=\"Motif\"\n",
    "    col3=\"Hits\"\n",
    "    tableHeader = '\\t'.join((col1.ljust(12), col2.ljust(8), col3.rjust(4))) \n",
    "    with open(f2, 'w') as output:\n",
    "        output.write(tableHeader + \"\\n\")\n",
    "        #scan seqOnly for motifs\n",
    "        x=1\n",
    "        y=0\n",
    "        motifHits = []\n",
    "        #define table pattern\n",
    "        for item in seqOnly:\n",
    "            motifList = re.findall(motif, item)\n",
    "            motifHits.append(len(motifList))\n",
    "            x=x+1\n",
    "        for item in namesOnly:\n",
    "            tablePattern = '\\t'.join((str(namesOnly[y]).ljust(4), motif.ljust(4), str(motifHits[y]).rjust(10))) + \"\\n\"\n",
    "            output.write(tablePattern)\n",
    "            y=y+1\n",
    "        output.close()\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
